[senshir] iP

senshir opened the pull request:
# Slipstream 🚀

Slipstream is an **efficient** task tracker at command line interface (CLI).

## What can you do? 🤔

- [X] Add, delete, and mark tasks as done
- [X] Find tasks using keywords
- [X] Automatically save and load tasks
- [ ] Remind you which tasks you should get done soon (upcoming)

Designed with efficiency and simplicity in mind, Slipstream provides a **smooth task-tracking** experience.

## How to use 🛠️

1. Clone this repository into your own:
   ```
   git clone https://github.com/senshir/slipstream.git
   ```

2. Navigate to the project directory:
   ```
   cd slipstream
   ```
> Note: Ensure you have Java 17/Azul JKD (for MacOS) installed before running the program

3. Compile and run the program:
   ```
   javac -d bin src/main/java/slipstream/Slipstream.java
   java -cp bin slipstream.Slipstream
   ```

## Usage 💻

* *list* - display all tasks
  *Example* - list
* *todo* - add a new todo task
  *Example* - todo Read a book
* *deadline* - add a task with a deadline  
  *Example* - deadline Submit report /by Monday  
* *event* - add an event with a specific time  
  *Example* - event Meeting /at 2pm  
* *mark* - mark a task as done  
  *Example* - mark 1  
* *unmark* - unmark a completed task  
  *Example* - unmark 1  
* *delete* - delete a task from the list  
  *Example* - delete 2  
* *find* - find tasks that contain a keyword  
  *Example* - find book  
* *bye* - exit Slipstream  
  *Example* - bye

## Additional Information
You may access a .jar file of Slipstream from this link [here](https://github.com/senshir/ip/releases/tag/A-Jar)


**Warning:** Keep the `src\main\java` folder as the root folder for Java files (i.e., don't rename those folders or move Java files to another folder outside of this folder path), as this is the default location some tools (e.g., Gradle) expect to find Java files.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

senshir committed:
Replaced Duke with Slipstream, and allowed it to repeat the inputs I give it

senshir committed:
Changed Slipstream's function from responding to what I say to adding tasks to a list

senshir committed:
Added a Task class to keep track of tasks, and allowed Slipstream to mark and unmark different tasks

senshir committed:
Level-4, using different classes to manage different types of tasks. Also included an automated text UI tester

senshir committed:
Added various exception handling using my custom class SlipstreamException. And, added a delete function for the bot. Tested using runtest and I think it works okay?

senshir committed:
Update Task.java and its subclasses

Changes made:
* Included a new constructor for an already done task (using boolean isDone)
* Included a new String conversion for saved files
* Made Task an abstract class

senshir committed:
Add FileSaving.java

FileSaving is a class that assists with:
* Saving tasks into a filepath
* Loading tasks from a filepath
* Reading tasks from a text file by converting from String in the text file to relevant Task type

senshir committed:
Update Slipstream.java

Changes made:
* Fixed the indentation with switch cases
* Created a FileSaving object and used it to load and save tasks in relevant commands
* Added a "clear" command for users to clear all tasks since deleting one by one may take too long

senshir committed:
Revert "Update Slipstream.java"

This reverts commit 23772724e1fb24f0d1d713e0ce0398006b493ab0.

senshir committed:
Reupdate Slipstream.java

Something went wrong with my merging and I thought reverting the commit would undo the merge,
but it didn't help. So I'm redoing the commit:

Changes made:
* Fixed the indentation with switch cases
* Created a FileSaving object and used it to load and save tasks in relevant commands
* Added a "clear" command for users to clear all tasks since deleting one by one may take too long

senshir committed:
Merge branch 'branch-Level-7'

senshir committed:
Update Deadline.java

Added a specific format for deadlines to follow
* A written format in yyyy-mm-dd will be reformatted to MMM dd yyyyy when listed
* A SlipstreamException is thrown if the format is written wrongly

senshir committed:
Update FileSaving.java and Slipstream.java

* SlipstreamException is now thrown in multiple methods when the date is written in the wrong format
--> Prevents a Deadline object from still being created

senshir committed:
Merge branch 'branch-Level-8'

senshir committed:
Rearranged all files

Modified the OOP of the code to make things more efficient:
* Renamed my initial FileSaving class to Storage as stated in the MoreOOP requirements
* Did use Github Copilot to autocomplete subsequent lines of switch cases in Parser and Ui messages to speed up codewriting

senshir committed:
Merge branch 'branch-A-MoreOOP'

senshir committed:
Added packages

Organised Slipstream into a main slipstream package with subpackages for:
* Ui
* Task
* Parser
* Storage

senshir committed:
Merge branch 'branch-A-Packages'

senshir committed:
Merge branch 'add-gradle-support'

senshir committed:
Add Gradle Build

Changes made with gradle:
* Many style changes using checkstyle
* Accidentally moved on to using JUnit before branching, so test classes are made

senshir committed:
Merge branch 'A-Gradle'

senshir committed:
Add SlipstreamTest and DeadlineTest for command validation

Slipstream currently lacks automated tests for handling valid and invalid user commands, including task creation and deadline formatting.
Without these tests, errors may go unnoticed, and regressions may occur.

Let's add,
* SlipstreamTest to verify valid and invalid commands, ensuring proper error handling.
* DeadlineTest to test correct and incorrect deadline formats, catching parsing issues early.

senshir committed:
Merge branch 'branch-A-JUnit'

senshir committed:
Set archiveFileName for shadowJar

Define archiveFileName as "slipstream.jar" to standardize output naming

senshir committed:
Merge branch 'branch-A-Jar'

senshir committed:
Add JavaDoc comments to Ui class

Added JavaDoc comments to all methods in the all classes to improve code readability and maintainability.
Each method now includes a description, parameter details, and return values where applicable.

senshir committed:
Merge branch 'branch-A-JavaDoc'

senshir committed:
CodingStandard: Remove unused files

The project was already mostly compliant with the coding standard due to earlier changes. No major modifications were needed.
This commit removes unused files to maintain a clean codebase.

senshir committed:
Merge branch 'branch-A-CodingStandard'

senshir committed:
Create find command to search tasks by keyword

Changes made are,
* Added findTasks method in TaskList to filter tasks by keyword.
* Updated Ui to display matching tasks.
* Modified Parser to handle find command.

senshir committed:
Merge branch 'branch-Level-9'

# Conflicts (resolved):
#	src/main/java/slipstream/task/Task.java

charlenelau-20 commented on src/main/java/slipstream/ui/Ui.java:
```diff
@@ -0,0 +1,148 @@
+        System.out.println("____________________________________________________________");
+        System.out.println(" Bye. Hope to see you again soon!");
+        System.out.println("____________________________________________________________");
+    }
```
Maybe making a method to print the line and using that method throughout would make the code cleaner?

charlenelau-20 commented on src/main/java/slipstream/parser/Parser.java:
```diff
@@ -0,0 +1,99 @@
+            }
+        } catch (Exception e) {
+            ui.showMessage("An error occurred: " + e.getMessage());
+        }
```
Perhaps can be more specific with what types of exceptions you are catching, as "Exception" is quite broad.

charlenelau-20 commented on src/main/java/slipstream/task/Deadline.java:
```diff
@@ -0,0 +1,77 @@
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+            this.by = LocalDate.parse(by, formatter);
+        } catch (DateTimeParseException e) {
+            throw new SlipstreamException("Please write your deadline in this format: yyyy-MM-dd,"
+                + " e.g., 2021-02-26");
```
Looks good, maybe just a minor layout fix: indentations for wrapped lines should be 8 spaces instead of 4! (saw this in other sections of code as well)

charlenelau-20 commented on src/main/java/slipstream/parser/Parser.java:
```diff
@@ -0,0 +1,99 @@
+ * The parser also handles exceptions and displays error messages to the user.
+ * Based on the command, the parser calls the appropriate methods in the {@code TaskList} and {@code Ui} classes.
+ */
+public class Parser {
+
```
I liked how the code in this class was nicely structured and easy to read!

charlenelau-20 reviewed (commented):
I found most of the code quite easy to read overall! Just minor fixes, and everything else looks pretty good to me!

senshir committed:
Checkstyle: update version and make minor tweaks

Updated the checkstyle file for better code readability.
Fixed minor issues such as adding missing Javadoc comments in various files.

senshir committed:
Merge branch 'branch-A-CheckStyle'

senshir committed:
GUI: implement chatbot interface with message display

Transformed the command line interface into a GUI that displays chatbot messages.
The GUI has been packaged for easier integration.

Future UX improvements are planned to enhance the visual appeal and user experience.

senshir committed:
Merge branch 'branch-Level-10'

senshir committed:
Parser & TaskList: Add assertions to minimise assumptions

Adding assertions to TaskList methods to ensure valid task indices
when marking, unmarking, or deleting tasks.

These assertions document the assumption that indices must be within
the valid range and help catch potential logic errors earlier.

senshir committed:
Parser: Improve code readability and maintainability

Refactored Parser to ensure methods are not too long:
* Switch cases were refactored into a separate method
* Created a handle method for every case
* Invalid command length methods were made

senshir committed:
Merge pull request #2 from senshir/branch-A-Assertions

Parser & TaskList: Add assertions to minimise assumptions

senshir committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/slipstream/parser/Parser.java

senshir committed:
Merge branch 'master' into branch-A-Streams

senshir committed:
Merge pull request #3 from senshir/branch-A-CodeQuality

Parser: Improve code readability and maintainability

senshir committed:
Merge branch 'master' into branch-A-Streams

senshir committed:
Merge pull request #4 from senshir/branch-A-Streams

Merge A-Streams into master

senshir committed:
Reschedule: Add reschedulability to deadline tasks

The new `reschedule` command allows users to change the deadline of
existing `Deadline` tasks by specifying the task index and a new date
in `yyyy-MM-dd` format.

Changes made:
* Implemented `rescheduleTask` method in `TaskList` to handle rescheduling.
  * Validates input format (`reschedule <task_number> <new_deadline>`).
  * Ensures the task number is valid and refers to a `Deadline` task.
  * Calls `setDeadline` to update the deadline, which checks date format.
  * Saves updated tasks to storage.
* Updated `Deadline` class to include `setDeadline` method for modifying deadlines.
* Added `showTaskRescheduled` method in `UiMessages` to display reschedule confirmation.

This improves task management by allowing users to postpone or adjust deadlines
without manually deleting and re-adding tasks.

senshir committed:
Merge branch 'branch-B-Reschedule'

senshir committed:
Tests: update SlipstreamTest to verify input-output accuracy

Refactored `SlipstreamTest.java` to align with the new chatbot interaction model.
The test now checks for output correctness.

Finalised Slipstream chatbot for submission.

senshir committed:
Merge branch 'branch-Finalising'

senshir committed:
Docs: update README with user guide and feature list

Expanded `README.md` to include a comprehensive user guide detailing Slipstream's features and usage.
This improves accessibility and helps users navigate the chatbot effectively.

senshir committed:
Merge branch 'branch-A-UserGuide'

senshir committed:
Docs: finalise user guide in README for submission

Fixed minor issues in the user guide to ensure clarity and completeness.
The README now fully documents Slipstream’s features and usage, marking it ready for submission.

senshir committed:
Merge branch 'branch-A-Release'
